src/Ractive/config/custom/data.js

```diff
@@ -18,82 +18,60 @@ var dataConfigurator = {
 	reset: function ( ractive ) {
 		var result = this.init( ractive.constructor, ractive, ractive.viewmodel );
 
-		if ( result ) {
-			ractive.viewmodel.data = result;
-			ractive.viewmodel.clearCache( '' );
-			return true;
-		}
+		ractive.viewmodel.reset( result );
+		return true;
 	}
 };
 
 export default dataConfigurator;
 
-function combine ( Parent, target, options ) {
-	var value = options.data,
-		parentValue = getAddedKeys( Parent.prototype.data );
-
-	if ( value && typeof value !== 'object' && typeof value !== 'function') {
-		throw new TypeError( 'data option must be an object or a function, "' + value + '" is not valid' );
+function combine ( parentValue, childValue ) {
+	if ( childValue && typeof childValue !== 'object' && typeof childValue !== 'function') {
+		throw new TypeError( 'data option must be an object or a function, "' + childValue + '" is not valid' );
 	}
 
 	// Very important, otherwise child instance can become
 	// the default data object on Ractive or a component.
 	// then ractive.set() ends up setting on the prototype!
-	if ( !value && typeof parentValue !== 'function' ) {
-		value = {};
+	if ( !childValue && typeof parentValue !== 'function' ) {
+		childValue = {};
 	}
 
-	return dispatch( parentValue, value );
-}
-
-function getAddedKeys ( parent ) {
-	// only for functions that had keys added
-	if ( typeof parent !== 'function' || !Object.keys( parent ).length ) { return parent; }
-
-	// copy the added keys to temp 'object', otherwise
-	// parent would be interpreted as 'function' by dispatch
-	let temp = {};
-	copy( parent, temp );
+	let parentIsFn = typeof parentValue === 'function';
+	let childIsFn = typeof childValue === 'function';
 
-	// roll in added keys
-	return dispatch( parent, temp );
-}
-
-function dispatch ( parent, child ) {
-	var parentIsFn = typeof parent === 'function',
-		childIsFn = typeof child === 'function';
-
-	if( !parentIsFn && !childIsFn ) {
-		return fromProperties( child, parent );
+	// Fast path, where we just need to copy properties from
+	// parent to child
+	if ( !parentIsFn && !childIsFn ) {
+		return fromProperties( childValue, parentValue );
 	}
 
-	return function(){
-		child = childIsFn ? child.call( this ) : child,
-		parent = parentIsFn ? parent.call(this) : parent;
+	return function () {
+		let child = childIsFn ? childValue.call( this ) : childValue;
+		let parent = parentIsFn ? parentValue.call( this ) : parentValue;
 
 		// allow parent return value to take precedence if
 		// it is a function that returns non-POJO Model
 		// and child is either not a function or does not return non-POJO
 		if ( ( parentIsFn && parent.constructor !== Object ) && ( !childIsFn || child.constructor === Object ) ) {
-			return fromProperties( parent, child, true );
```

Comment:
If you take the force out, then the child keys won't override the model keys:

```js

function Model(){
    this.foo = 'qux';
}
var View = Ractive.extend({
    data: function(){
        return new Model();
    }
});

var ractive = new View({
    data: {
        foo: 'bar'
    }
});

// will only be 'bar' if child keys are forced onto parent
// because the retained object has been switched to the Model
ractive.get('foo') === 'bar'
```